clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubBasedAdminClientManager.java

```diff
@@ -114,74 +103,80 @@ private Future<SubscriptionAdminClient> getOrCreateSubscriptionAdminClient() {
+    private Future<String> createTopic(final String projectId, final TopicName topic, final TopicAdminClient client) {
         try {
             final Topic createdTopic = client.createTopic(topic);
             if (createdTopic == null) {
                 LOG.debug("Creating topic failed [topic: {}, projectId: {}]", topic, projectId);
                 return Future.failedFuture("Topic creation failed.");
             }
-            topics.add(createdTopic.getName());
-            return Future.succeededFuture(topic.getTopic());
-        } catch (AlreadyExistsException ex) {
+            LOG.debug("Topic {} created successfully.", topic);
             return Future.succeededFuture(topic.getTopic());
         } catch (ApiException e) {
             LOG.debug("Error creating topic {} on project {}", topic, projectId, e);
             return Future.failedFuture("Topic creation failed.");
         }
-
     }
 
     /**
-     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and prefix.
+     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and
+     * prefix.
      *
      * @param endpoint The endpoint name of the subscription, e.g. command_internal.
      * @param prefix The prefix of the subscription, e.g. the adapter instance ID.
-     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed
-     *         Future if it could not be created.
+     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed Future if
+     *         it could not be created.
      */
     public Future<String> getOrCreateSubscription(final String endpoint, final String prefix) {
         final String topicAndSubscriptionName = PubSubMessageHelper.getTopicName(endpoint, prefix);
         final TopicName topic = TopicName.of(projectId, topicAndSubscriptionName);
         final SubscriptionName subscription = SubscriptionName.of(projectId, topicAndSubscriptionName);
 
-        if (subscriptions.contains(subscription.toString())) {
-            LOG.debug("Subscription {} already exists, continue", subscription);
-            return Future.succeededFuture(subscription.getSubscription());
-        }
         return getOrCreateSubscriptionAdminClient()
                 .onFailure(thr -> LOG.debug("admin client creation failed", thr))
-                .compose(client -> getOrCreateSubscription(projectId, subscription, topic, client));
+                .compose(client -> getSubscription(subscription, client)
+                        .recover(thr -> createSubscription(projectId, subscription, topic, client)));
+    }
+
+    private Future<String> getSubscription(final SubscriptionName subscription, final SubscriptionAdminClient client) {
+        try {
+            return Future.succeededFuture(client.getSubscription(subscription).getName());
```

Comment:
I am not an expert regarding the PubSub client but I wonder if this (blocking) call might take some time to execute, thus blocking the event loop?

Reply:
I am not sure either, but I imagine that this could take a bit of time in some scenarios. That being said, I have never seen a BlockedThreadChecker log saying that this was blocking the event loop longer than 2 seconds in our case. Maybe we are on the saver side though if I add a comment to the JavaDoc stating that this call is blocking. What do you think?

Reply:
Well, given that it doesn't take a handler nor returns a future, it seems to be clear that this is a blocking call. If we want to be on the safe side, then I guess we should wrap it in a `Context.executeBlocking()` invocation ...

Reply:
The problem with wrapping it in a `Context.executeBlocking()` invocation is that we use this to create the internal command topics and subscriptions upon starting the adapter instances. I think having this as non-blocking in the startup process could lead to problems when devices start connecting to the adapter before the topic and subscription are in place.

Reply:
Well, if you really want to make this method being usable in a blocking manner only, then the method should return a plain `String` instead of a `Future<String>`.
However, FMPOV it makes more sense to return a Future and let the invoking code in the adapter deal with the fact that it needs to wait for the returned future to be completed before marking the adapter as *operational/up/ready*.

Reply:
> However, FMPOV it makes more sense to return a Future and let the invoking code in the adapter deal with the fact that it needs to wait for the returned future to be completed before marking the adapter as _operational/up/ready_.

I agree, but I am a bit confused now as this function already returns a future. Could you maybe clarify a bit more how your preferred solution would look like?

Reply:
Return a `Future<String>` and run the code using `Context.executeBlocking()`.

Reply:
Done. I'm sorry it took so long; other things got in the way.

Reply:
No worries :-) This is an open source project and we all spend as much time as we are willing to or can afford. It might take some time but we usually get better results ...